src/psyclone/psyir/nodes/dynamic_omp_task_directive.py

```diff
@@ -0,0 +1,1998 @@
+        :param shared_list: The list of shared References for this task.
+        :type shared_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+        :param out_list: The list of output References for this task.
+        :type out_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+        """
+        # Check if its a private variable
+        is_private = self._is_reference_private(ref)
+        # If its private should add it to private list if not already present
+        if is_private and ref not in private_list:
+            private_list.append(ref.copy())
+        # Otherwise its a shared variable
+        if not is_private:
+            if ref not in shared_list:
+                shared_list.append(ref.copy())
+            if ref not in out_list:
+                out_list.append(ref.copy())
+
+    def _evaluate_write_reference(
+        self, ref, private_list, firstprivate_list, shared_list, out_list
+    ):
+        """
+        Evaluates a write to any Reference in the task region. This is done by
+        calling the appropriate subfunction depending on the type of the
+        Reference.
+
+        :param ref: The Reference to be evaluated.
+        :type ref: :py:class:`psyclone.psyir.nodes.Reference`
+        :param private_list: The list of private References used in this task
+                             region.
+        :type private_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+        :param firstprivate_list: The list of firstprivate References used in
+                                  this task region.
+        :type firstprivate_list: List of
+                                 :py:class:`psyclone.psyir.nodes.Reference`
+        :param shared_list: The list of shared References for this task.
+        :type shared_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+        :param out_list: The list of output References for this task.
+        :type out_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+
+        :raises GenerationError: If a StructureReference containing multiple \
+                                 ArrayMember or ArrayOfStructuresMember as \
+                                 children is found.
+        """
+        if isinstance(ref, (ArrayReference, ArrayOfStructuresReference)):
+            self._evaluate_write_arrayref(
+                ref, private_list, firstprivate_list, shared_list, out_list
+            )
+        elif isinstance(ref, StructureReference):
+            # If the StructureReference contains an ArrayMixin then
+            # we need to treat it differently, like an arrayref, however
+            # the code to handle an arrayref is not compatible with more
+            # than one ArrayMixin child
+            array_children = ref.walk((ArrayOfStructuresMember, ArrayMember))
+            if len(array_children) > 0:
+                if len(array_children) > 1:
+                    raise GenerationError(
+                        "Doesn't support a "
```

Comment:
Please reword slightly and include the reference expression in the message.

Reply:
Done